pkg/controller/podautoscaler/metrics/metrics_client.go

```diff
@@ -42,292 +62,169 @@ const (
 	params := map[string]string{"labelSelector": selector.String()}
 
-	resultRaw, err := h.client.Core().Services(h.heapsterNamespace).
+	resultRaw, err := h.services.
 		ProxyGet(h.heapsterScheme, h.heapsterService, h.heapsterPort, metricPath, params).
 		DoRaw()
 	if err != nil {
-		return 0, time.Time{}, fmt.Errorf("failed to get pods metrics: %v", err)
+		return nil, time.Time{}, fmt.Errorf("failed to get pod resource metrics: %v", err)
 	}
 
 	glog.V(4).Infof("Heapster metrics result: %s", string(resultRaw))
 
 	metrics := metrics_api.PodMetricsList{}
 	err = json.Unmarshal(resultRaw, &metrics)
 	if err != nil {
-		return 0, time.Time{}, fmt.Errorf("failed to unmarshall heapster response: %v", err)
+		return nil, time.Time{}, fmt.Errorf("failed to unmarshal heapster response: %v", err)
 	}
 
-	if len(metrics.Items) != len(podNames) {
-		present := sets.NewString()
-		for _, m := range metrics.Items {
-			present.Insert(m.Name)
-		}
-		missing := make([]string, 0)
-		for expected := range podNames {
-			if !present.Has(expected) {
-				missing = append(missing, expected)
-			}
-		}
-		hint := ""
-		if len(missing) > 0 {
-			hint = fmt.Sprintf(" (sample missing pod: %s/%s)", namespace, missing[0])
-		}
-		return 0, time.Time{}, fmt.Errorf("metrics obtained for %d/%d of pods%s", len(metrics.Items), len(podNames), hint)
+	if len(metrics.Items) == 0 {
+		return nil, time.Time{}, fmt.Errorf("no metrics returned from heapster")
 	}
 
-	sum := int64(0)
+	res := make(PodResourceInfo, len(metrics.Items))
+
 	for _, m := range metrics.Items {
-		if _, found := podNames[m.Name]; found {
-			for _, c := range m.Containers {
-				cpu, found := c.Usage[v1.ResourceCPU]
-				if !found {
-					return 0, time.Time{}, fmt.Errorf("no cpu for container %v in pod %v/%v", c.Name, namespace, m.Name)
-				}
-				sum += cpu.MilliValue()
+		podSum := int64(0)
+		missing := len(m.Containers) == 0
+		for _, c := range m.Containers {
+			cpu, found := c.Usage[v1.ResourceName(resource)]
+			if !found {
+				missing = true
+				glog.V(2).Infof("missing resource metric %v for container %s in pod %s/%s", resource, c.Name, namespace, m.Name)
+				continue
 			}
-		} else {
-			return 0, time.Time{}, fmt.Errorf("not expected metrics for pod %v/%v", namespace, m.Name)
+			podSum += cpu.MilliValue()
 		}
-	}
 
-	return sum / int64(len(metrics.Items)), metrics.Items[0].Timestamp.Time, nil
-}
+		if !missing {
+			res[m.Name] = int64(podSum)
+		}
+	}
 
-// GetCustomMetric returns the average value of the given custom metric from the
-// pods picked using the namespace and selector passed as arguments.
-func (h *HeapsterMetricsClient) GetCustomMetric(customMetricName string, namespace string, selector labels.Selector) (*float64, time.Time, error) {
-	metricSpec := getHeapsterCustomMetricDefinition(customMetricName)
+	timestamp := time.Time{}
+	if len(metrics.Items) > 0 {
+		timestamp = metrics.Items[0].Timestamp.Time
```

Comment:
It's what we do now: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/metrics/metrics_client.go#L208